operator/api/v1alpha2/seldondeployment_types.go

```diff
@@ -0,0 +1,372 @@
+		if len(svcName) > 63 {
+			return "seldon-" + hash(svcName)
+		} else {
+			return svcName
+		}
+	} else {
+		return svcName
+	}
+}
+
+// EDIT THIS FILE!  THIS IS SCAFFOLDING FOR YOU TO OWN!
+// NOTE: json tags are required.  Any new fields you add must have json tags for the fields to be serialized.
+
+// SeldonDeploymentSpec defines the desired state of SeldonDeployment
+type SeldonDeploymentSpec struct {
+	Name        string            `json:"name,omitempty" protobuf:"string,1,opt,name=name"`
+	Predictors  []PredictorSpec   `json:"predictors,omitempty" protobuf:"bytes,2,opt,name=name"`
+	OauthKey    string            `json:"oauth_key,omitempty" protobuf:"string,3,opt,name=oauth_key"`
+	OauthSecret string            `json:"oauth_secret,omitempty" protobuf:"string,4,opt,name=oauth_secret"`
+	Annotations map[string]string `json:"annotations,omitempty" protobuf:"bytes,5,opt,name=annotations"`
+}
+
+type PredictorSpec struct {
+	Name            string                  `json:"name,omitempty" protobuf:"string,1,opt,name=name"`
+	Graph           *PredictiveUnit         `json:"graph,omitempty" protobuf:"bytes,2,opt,name=predictiveUnit"`
+	ComponentSpecs  []*SeldonPodSpec        `json:"componentSpecs,omitempty" protobuf:"bytes,3,opt,name=componentSpecs"`
+	Replicas        int32                   `json:"replicas,omitempty" protobuf:"string,4,opt,name=replicas"`
+	Annotations     map[string]string       `json:"annotations,omitempty" protobuf:"bytes,5,opt,name=annotations"`
+	EngineResources v1.ResourceRequirements `json:"engineResources,omitempty" protobuf:"bytes,6,opt,name=engineResources"`
+	Labels          map[string]string       `json:"labels,omitempty" protobuf:"bytes,7,opt,name=labels"`
+	SvcOrchSpec     SvcOrchSpec             `json:"svcOrchSpec,omitempty" protobuf:"bytes,8,opt,name=svcOrchSpec"`
+	Traffic         int32                   `json:"traffic,omitempty" protobuf:"bytes,9,opt,name=traffic"`
+	Explainer       Explainer               `json:"explainer,omitempty" protobuf:"bytes,10,opt,name=explainer"`
+}
+
+type SvcOrchSpec struct {
+	Resources *v1.ResourceRequirements `json:"resources,omitempty" protobuf:"bytes,1,opt,name=resources"`
+	Env       []*v1.EnvVar             `json:"env,omitempty" protobuf:"bytes,2,opt,name=env"`
+}
+
+type Explainer struct {
+	Type               string            `json:"type,omitempty" protobuf:"string,1,opt,name=type"`
+	ModelUri           string            `json:"modelUri,omitempty" protobuf:"string,2,opt,name=modelUri"`
+	ServiceAccountName string            `json:"serviceAccountName,omitempty" protobuf:"string,3,opt,name=serviceAccountName"`
+	ContainerSpec      v1.Container      `json:"containerSpec,omitempty" protobuf:"bytes,4,opt,name=containerSpec"`
+	Config             map[string]string `json:"config,omitempty" protobuf:"bytes,5,opt,name=config"`
+	Endpoint           *Endpoint         `json:"endpoint,omitempty" protobuf:"bytes,6,opt,name=endpoint"`
```

Comment:
Endpoint is pretty small?  why using reference?

Reply:
its optional. We don't want an empy struct.